### PR TITLE
storage: avoid reflection-based transaction cloning

### LINKED ISSUE #10868: storage: panic cloning roachpb.Error during lease acquisition
Saw this panic on `delta` while it was running #10867.

```
panic: attempt to clone range 72 was not found, which contains 0d5e20b0-fb94-4fb2-ab9c-eae7c7099e15 [recovered]
```

The `range 72 was not found` sounds like a `roachp.RangeNotFoundError`. Not sure how it could contain that UUID. The stack trace is below.

```
panic(0x16f07e0, 0xc449552680)
	/usr/local/go/src/runtime/panic.go:458 +0x243
github.com/cockroachdb/cockroach/pkg/util/protoutil.Clone(0x23fd860, 0xc42ba323c0, 0xc4233a68d8, 0x1488d4772c2167f9)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/protoutil/clone.go:60 +0x247
github.com/cockroachdb/cockroach/pkg/storage.(*pendingLeaseRequest).InitOrJoinRequest.func1(0x7f5b2dff3f80, 0xc42e98e810)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica_range_lease.go:154 +0x359
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask.func1(0xc420096000, 0x20eb9dc, 0x1e, 0x9f, 0x0, 0x0, 0xc42e98e720, 0x7f5b2dfe8148, 0xc42000e280)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:264 +0xdf
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:265 +0x234
```

Cc @tamird who knows the proto cloning code and @andreimatei who knows the lease code

### PR DESCRIPTION
**Pull request by tamird**
Fixes #10868.

---
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/10869)

### CODE REVIEW

File: pkg/storage/replica_range_lease.go [hunk 1]
```diff
@@ -143,15 +142,14 @@ func (p *pendingLeaseRequest) InitOrJoinRequest(
 		// Send result of lease to all waiter channels.
 		replica.mu.Lock()
 		defer replica.mu.Unlock()
-		for i, llChan := range p.llChans {
-			// Don't send the same pErr object twice; this can lead to races. We could
-			// clone every time but it's more efficient to send pErr itself to one of
-			// the channels (the last one; if we send it earlier the race can still
-			// happen).
-			if i == len(p.llChans)-1 {
```
Comment (petermattis) [inline pkg/storage/replica_range_lease.go]: This was only cloning the error when sending to `p.llChans[len(p.llChans)-1]`. Now you're always the error if it is not `nil`. Probably not a big deal, but I'd prefer to be conservative in this change.

Comment (tamird) [inline pkg/storage/replica_range_lease.go]: You mean it was _not_ cloning the error in that case? Yes, that's true, and yet cloning the error every time is the more conservative behaviour.

Comment (tamird) [inline pkg/storage/replica_range_lease.go]: OK, done.

**Review by petermattis — APPROVED** [review]
LGTM

**Comment by andreimatei** [comment]
LGTM, thanks for the fix

Would you mind adding a test? :) The thing to test, I guess, would be two guys waiting for the result of a lease request that results in an error.
I think you can copy the beginning of `TestBehaviorDuringLeaseTransfer` and do a second get while a transfer is blocked.

---

Review status: 0 of 1 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10869#-:-KX2N0U-p_dSaPhXcbqv:b-yjb22z)*

**Comment by tamird** [comment]
How do I get the returned error to include a transaction?

**Comment by andreimatei** [comment]
I hope you can just construct whatever `pErr` you want inside the `TestingCommandFilter`. Chose one that has a transaction proto in it.
If you want the test to actually assert something, rather than just not crashing, instead of `Get`s, you could send multiple concurrent `AdminTransferLease` with the same target, and inspect the errors they get.

File: pkg/storage/replica_range_lease.go (hunk 1 above)
Comment (petermattis) [inline pkg/storage/replica_range_lease.go]: Ok, disregard my comment and change this back to always cloning. Not cloning seems to cause concurrent map access. Also, in the `pErr == nil` case, I think it is clearer to send on the channel using `llChan <- err`.

**Review by petermattis — APPROVED** [review]
LGTM

Unless a test is imminent, let's get this merged (and quickly follow-up with a test) as it is happening in test clusters.

File: pkg/storage/replica_range_lease.go (hunk 1 above)
Comment (petermattis) [inline pkg/storage/replica_range_lease.go]: Gah, make that `llChan <- nil`.

**Comment by bdarnell** [comment]
How does a range not found error for a lease request have a non-nil transaction?

We should definitely have a test that covers this error-copying scenario, but it's also error-prone that the specific error value matters. I think protoutil.Clone should be a static check based on the type of the protobuf instead of a dynamic check that depends on which fields are non-empty. (there are some other bugs in protoutil too; PR coming shortly)

---

Review status: 0 of 1 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10869#-:-KX3pzRRmjQHQjiYd98l:b-owicec)*

File: pkg/storage/replica_range_lease.go (hunk 1 above)
Comment (tamird) [inline pkg/storage/replica_range_lease.go]: OK, done again.

**Comment by tamird** [comment]
@bdarnell I agree that it'd be good to understand how we get a range not found error with a transaction - currently, I have no idea.

A static check in protoutil.Clone would be nice, but it's not always feasible IIRC. I'll look for your PR.

**Comment by bdarnell** [comment]
FYI TestLeaseConcurrent is our current test for cloning the error. 

---

Review status: 0 of 1 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10869#-:-KX40GOq2LXnZ6N2Yxdz:baz0lvl)*

**Comment by bdarnell** [comment]
:lgtm:

---

Review status: 0 of 1 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10869#-:-KX419660drJc5ggSSpa:bnfp4nl)*

**Pull request merged**